h/util/logging_filters.py

```diff
@@ -0,0 +1,53 @@
+# -*- coding: utf-8 -*-
+"""Logging Filters."""
+
+import logging
+
+
+# logging levels from https://docs.python.org/2/library/logging.html#logging-levels
+LEVELS = {"CRITICAL": 50,
+          "ERROR": 40,
+          "WARNING": 30,
+          "INFO": 20,
+          "DEBUG": 10,
+          "NOTSET": 0}
+
+
+class ExceptionFilter(logging.Filter):
+    """Filter out the specified exceptions with specified logging level."""
+
+    def __init__(self, ignore_exceptions):
+        """
+        Configure filtering out of the specified exceptions with specified logging level.
+
+        ignore_exceptions: a tuple of tuples ((exception type, loglevel))
+                           example: (("requests.exceptions.ReadTimeout", "WARNING"),)
+        """
+        super(ExceptionFilter, self).__init__()
+
+        # build decoder dict where both string and int map to int logging level
+        logging_levels = {val: val for val in LEVELS.values()}
+        logging_levels.update(LEVELS)
+
+        self._ignore_exceptions = []
+        for (exc_type, exc_level) in ignore_exceptions:
+            exception_idx = exc_type.rfind('.')
+            module, exception = exc_type[:exception_idx], exc_type[exception_idx+1:]
+            try:
+                self._ignore_exceptions.append((logging_levels[exc_level],
+                                                getattr(__import__(module), exception)))
```

Comment:
While testing this locally, I found a subtle issue with `__import__` which causes this function not to behave as expected if `module` is a dotted path, for example if `exc_type` is `"urllib3.exceptions.ReadTimeoutError"`, then. `__import__('urllib3.exceptions')` returns a reference to the top-level module ("urllib3") and so the `getattr` call may fail.

[From the docs](https://docs.python.org/2/library/functions.html#__import__):

> When the name variable is of the form package.module, normally, the top-level package (the name up till the first dot) is returned, not the module named by name. However, when a non-empty fromlist argument is given, the module named by name is returned.

This code is however working for `requests.exceptions.ReadTimeout`. Why? Because the `requests` package re-exports the `ReadTimeout` class at the top level so `from requests import ReadTimeout` works.

Reply:
Wah! Good catch! I fix...

Reply:
So this converting a string path to the exception class is becoming overly complicated. Are we ok with just accepting the exception name rather than the full path? That will simplify the code dramatically. (I'm committing it with this change but I can undo it if you don't like it.) One advantage of using just the name would be it would handle differences between the dev and prod env (because dev uses urllib3 and prod uses http_requests). Alternatively we can just change it to not be a string as you suggested earlier.

Reply:
> Are we ok with just accepting the exception name rather than the full path? 

While that's probably Good Enough for our needs right now, I would still be inclined to avoid the string conversion and just require the exception class to be passed, until we have a concrete need to specify the exception type as a string.

If that concrete need does ever arise, an option I forgot about yesterday is that Pyramid exposes a [utility class](https://docs.pylonsproject.org/projects/pyramid/en/latest/api/path.html#pyramid.path.DottedNameResolver) which can resolve dotted names to classes. See `h/cli/__init__.py` for example usage.

> One advantage of using just the name would be it would handle differences between the dev and prod env (because dev uses urllib3 and prod uses http_requests). 

The exception has a slightly different name in `requests` and `urllib3` - `ReadTimeout` vs `ReadTimeoutError`.